File: ex_happy-primes/submissions/roberto_bonvallet-py/happyprime.py
import re
s=lambda n,l:0if n==1 else n in l or s(sum(int(a)**2for a in `n`),l+[n])
n=input()
print['happy','sad'][s(n,[])],'non-'*bool(re.match(r'1?$|(11+?)\1+$','1'*n))+'prime'

File: ex_happy-primes/submissions/sepl-pl/happyprime.pl
sub h{my$s;$s+=$_**2for split//,pop;($s-4)?($s-1)?&h($s):1:0}$a=pop;
print h($a)?happy:sad,$",(1x$a)=~/^1?$|^(11+?)\1+$/&&"non-",prime
